Update dev container content

bamurtaugh opened the pull request:
Based on team discussions and customer feedback, we've been assembling a list of docs to update: https://github.com/microsoft/vscode-docs/issues/5737.

Summary of changes:
- Revisit Advanced Containers, removing one topic and adding a new one (help reduce the length of containers.md by extracting the git credential section into a new advanced page)
- Update based on latest spec and CLI updates, terminology, and guidance 
     - Major one was updating devcontainer.json examples to follow the latest spec and point to the latest images
- Changes based on customer feedback, including content that was outdated or wording/info that was confusing
- Separately, I updated aka.ms links along with CLI output info in containers.dev and devcontainers/cli
- As the last remaining item in issue 5737, I'll also separately update our dev container Learn module after this

samruddhikhandale commented on docs/devcontainers/create-dev-container.md:
````diff
@@ -184,7 +184,7 @@ When you make changes like installing new software, changes made in the Dockerfi
 
 ```docker
-FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:0-12
+FROM mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye
````
```suggestion
FROM mcr.microsoft.com/devcontainers/javascript-node:0-18
```

`0-18` and `0-18-bullseye` correlate to the same image. To keep things simple, should we avoid using `-bullseye` or specific OS (wherever possible) ?

bamurtaugh replied:
Makes sense to me!

samruddhikhandale commented on docs/devcontainers/tutorial.md:
```diff
@@ -114,53 +114,69 @@ This next section describes in more detail how the Dev Containers extension sets
+    "name": "Node.js",
+
+    // Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye",
```
```suggestion
	"image": "mcr.microsoft.com/devcontainers/javascript-node:0-18",
```

Same comment as above

samruddhikhandale commented on docs/remote/ssh.md:
```diff
@@ -46,7 +46,7 @@ To get started, you need to:
 2. Install [Visual Studio Code](https://code.visualstudio.com/) or [Visual Studio Code Insiders](https://code.visualstudio.com/insiders/).
 
-3. Install the [Remote Development extension pack](https://aka.ms/vscode-remote/download/extension).
+3.  Install the [Remote-SSH extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh). If you plan to work with other remote extensions in VS Code, you may choose to install the [Remote Development extension pack](https://aka.ms/vscode-remote/download/extension).
```
```suggestion
3. Install the [Remote-SSH extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-ssh). If you plan to work with other remote extensions in VS Code, you may choose to install the [Remote Development extension pack](https://aka.ms/vscode-remote/download/extension).
```

samruddhikhandale commented on remote/advancedcontainers/start-processes.md:
````diff
@@ -66,7 +66,7 @@ The `overrideCommand` property defaults to `true` because many images will immed
 
 ```docker
-FROM mcr.microsoft.com/vscode/devcontainers/base:0-focal
+FROM mcr.microsoft.com/devcontainers/base:ubuntu
````
```suggestion
FROM mcr.microsoft.com/devcontainers/base:1-ubuntu
```

samruddhikhandale reviewed (commented):
Looks great, left minor comments!

bamurtaugh commented:
Thanks so much for the fast review, @samruddhikhandale! Merged all your suggestions.

samruddhikhandale commented on docs/devcontainers/tutorial.md:
```diff
@@ -114,53 +114,69 @@ This next section describes in more detail how the Dev Containers extension sets
 The Dev Containers extension uses the files in the `.devcontainer` folder, namely `devcontainer.json`, and an optional `Dockerfile` or `docker-compose.yml`, to create your dev containers.
 
-First your image is built from the supplied Docker file or image name. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`.
+In the example we just explored, the project has a `.devcontainer` folder with a `devcontainer.json` inside. The `devcontainer.json` uses the image `mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye`. You can explore this image in greater detail in the [devcontainers/images](https://github.com/devcontainers/images/tree/main/src/javascript-node) repo.
```
```suggestion
In the example we just explored, the project has a `.devcontainer` folder with a `devcontainer.json` inside. The `devcontainer.json` uses the image `mcr.microsoft.com/devcontainers/javascript-node:0-18`. You can explore this image in greater detail in the [devcontainers/images](https://github.com/devcontainers/images/tree/main/src/javascript-node) repo.
```

samruddhikhandale commented on docs/devcontainers/tutorial.md:
```diff
@@ -114,53 +114,69 @@ This next section describes in more detail how the Dev Containers extension sets
-First your image is built from the supplied Docker file or image name. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`.
+In the example we just explored, the project has a `.devcontainer` folder with a `devcontainer.json` inside. The `devcontainer.json` uses the image `mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye`. You can explore this image in greater detail in the [devcontainers/images](https://github.com/devcontainers/images/tree/main/src/javascript-node) repo.
+
+First, your image is built from the supplied Dockerfile or image name, which would be `mcr.microsoft.com/devcontainers/javascript-node:0-18-bullseye` in this example. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`. For examp ...
```
```suggestion
First, your image is built from the supplied Dockerfile or image name, which would be `mcr.microsoft.com/devcontainers/javascript-node:0-18` in this example. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`. For example, the dev container in this example installs the `streetsidesoftware.code-spell-checker` extension.
```

samruddhikhandale reviewed (commented):
Thanks! Missed `-bullseye` in two other places.

bamurtaugh commented:
Thanks! Just merged those changes in.

Chuxel commented on docs/devcontainers/containers.md:
```diff
@@ -238,11 +238,11 @@ This implies trusting [the machine the Docker daemon runs on](/remote/advancedco
 VS Code's container configuration is stored in a [devcontainer.json](https://containers.dev/implementors/json_reference) file. This file is similar to the `launch.json` file for debugging configurations, but is used for launching (or attaching to) your development container instead. You can also specify any extensions to install once the container is running or post-create commands to prepare the ...
 
-You can use any image, Dockerfile, or set of Docker Compose files as a starting point. Here is a simple example that uses one of the pre-built [Development Container images](https://github.com/devcontainers/images):
+You can use any image, Dockerfile, or set of Docker Compose files as a starting point. Here is a simple example that uses one of the pre-built [Development Container images](https://github.com/devcontainers/images/tree/main/src/typescript-node):
```
I wonder if we want to point this to an index of the images. We don't have an index in MCR yet, but I suppose we could point at the tracking issue for now: https://github.com/devcontainers/images/issues/90

bamurtaugh replied:
I'd consider the devcontainers/images repo a comprehensive index for users trying to explore what images we offer. Is there a limitation to the /images repo that maybe I'm not considering?

Chuxel replied:
Yeah it doesn't help with the actual image names - you end up having to look at each README to see it. That may be fine for now tho.

bamurtaugh replied:
I think it's fine for now, and avoids users needing to read through and understand an additional issue. I also think it's not the total worst to point them to where they can learn more about each image (by exploring each README), so I'll leave this as-is at least for now.

Chuxel commented on docs/devcontainers/containers.md:
````diff
@@ -238,11 +238,11 @@ This implies trusting [the machine the Docker daemon runs on](/remote/advancedco
 ```json
 {
-    "image": "mcr.microsoft.com/devcontainers/typescript-node:0-12",
+    "image": "mcr.microsoft.com/devcontainers/typescript-node",
````
One other thing I wonder about this example, is that the eslint extension will already be installed by referencing typescript-node.

We could call that out, and add something else - in the sample I used the spell checker extension and then pointed out the two behaviors under "Edit".

bamurtaugh replied:
Updated to using the spellchecker extension here too, and added a note underneath about inheriting any extensions from the base image.

bamurtaugh replied:
Updated with a link to the pre-building section, and added info on labels to the pre-building section!

Chuxel commented on docs/devcontainers/containers.md:
```diff
@@ -311,9 +311,9 @@ We recommend pre-building images with the tools you need rather than creating an
 Even better - pre-built images can contain Dev Container metadata so when you reference an image, settings will be pulled across automatically.
 
-We recommend using the [devcontainer CLI](/docs/devcontainers/devcontainer-cli.md) to pre-build your images since it is kept in sync with the Dev Containers extension's latest capabilities - including [dev container Features](#dev-container-features). Once you've built your image, you can push it to a container registry (like the [Azure Container Registry](https://learn.microsoft.com/azure/contai ...
+We recommend using the [dev container CLI](/docs/devcontainers/devcontainer-cli.md) to pre-build your images since it is kept in sync with the Dev Containers extension's latest capabilities - including [dev container Features](#dev-container-features). Once you've built your image, you can push it to a container registry (like the [Azure Container Registry](https://learn.microsoft.com/azure/conta ...
```
Maybe we should point out the GitHub Action too?

bamurtaugh replied:
Makes sense, added a link to devcontainers/ci toward the end of this section.

Chuxel commented on docs/devcontainers/create-dev-container.md:
````diff
@@ -49,7 +49,7 @@ For this example, if you'd like to install the [ESLint extension](https://market
 ```json
 {
-    "image": "mcr.microsoft.com/devcontainers/typescript-node:0-12",
+    "image": "mcr.microsoft.com/devcontainers/typescript-node:0-18",
````
Same comment here - it might be worth highlighting image metadata here and using a different extension other than eslint.

bamurtaugh replied:
Just made the same updates here!

bamurtaugh replied:
Also linking to the more detailed pre-build and label info here now.

Chuxel commented on docs/devcontainers/devcontainer-cli.md:
```diff
@@ -29,24 +29,15 @@ When tools like VS Code and Codespaces detect a `devcontainer.json` file in a us
 ## System requirements
 
-To use the VS Code dev container CLI, you'll need the following on your system or CI/DevOps environment:
+You can quickly try out the CLI in just a few simple steps, either by installing its npm package or building the CLI repo from sources.
```
Given VS Code now ships a pre-built version of the CLI that also adds the "open" command, I wonder if we shouldn't push people towards using the "Install Dev Container CLI" command first.  We can then point out that there's other options for using the CLI elsewhere:

1. Install via npm - which then has these requirements right now.
2. Use the GitHub Action or Azure DevOps Task

bamurtaugh replied:
Added info on using the install command from the Dev Containers extension, and restructured a bit to add the Action and ADO task.

Chuxel commented on docs/devcontainers/devcontainer-cli.md:
```diff
@@ -64,6 +55,8 @@ Commands:
   devcontainer build [path]         Build a dev container image
   devcontainer run-user-commands    Run user commands
   devcontainer read-configuration   Read configuration
+  devcontainer features             Features commands
```
We can probably highlight "open" as an option when installed via VS Code.

bamurtaugh replied:
Added a note on this.

Chuxel commented:
In the CLI section, we still tell people to add their features again to the secondary devcontainer.json.  That's no longer required in the image pre-build situation so.

> 1. Create a simplified `devcontainer.json` file in repositories where you'd like to use the image - the `devcontainer.json` should either use the `image` property or reference the image in an associated Docker Compose file. For example:
>
>    ```json
>    {
>        "image": "ghcr.io/your-org/your-image-name"
>    }
>    ```

Chuxel commented on docs/devcontainers/tutorial.md:
```diff
@@ -114,53 +114,69 @@ This next section describes in more detail how the Dev Containers extension sets
-First your image is built from the supplied Docker file or image name. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`.
+In the example we just explored, the project has a `.devcontainer` folder with a `devcontainer.json` inside. The `devcontainer.json` uses the image `mcr.microsoft.com/devcontainers/javascript-node:0-18`. You can explore this image in greater detail in the [devcontainers/images](https://github.com/devcontainers/images/tree/main/src/javascript-node) repo.
+
+First, your image is built from the supplied Dockerfile or image name, which would be `mcr.microsoft.com/devcontainers/javascript-node:0-18` in this example. Then a container is created and started using some of the settings in the `devcontainer.json`. Finally your Visual Studio Code environment is installed and configured again according to settings in the `devcontainer.json`. For example, the d ...
```
This is probably another good spot to highlight that this image has metadata pre-built into it... the result would be that both eslint and the spellchecker extension would be installed.

bamurtaugh replied:
Makes sense, added that same note.

bamurtaugh replied:
Also updated this note with the link to pre-build info.

Chuxel commented:
I almost wonder if we should just remove the `cacheFrom` bit from the CLI section. I continue to see challenges with that and it may be just easier to keep things consistent.

bamurtaugh commented:
@Chuxel thanks so much for the detailed review! I believe I've addressed and replied to your comments above.

Chuxel commented:
One thing I'm not sure we have a good link to anywhere is image metadata. We allude to it in the pre-builds section but we're a bit more specific in the containers.dev docs.

https://containers.dev/implementors/reference/#labels

bamurtaugh commented:
Thanks @Chuxel! I expanded the info on metadata in the pre-build section (like what you added in the linked containers.dev PR), and I added links to this content from the **Notes** I had added on metadata throughout other docs.

Chuxel commented on docs/devcontainers/containers.md:
```diff
@@ -311,9 +311,9 @@ We recommend pre-building images with the tools you need rather than creating an
 Even better - pre-built images can contain Dev Container metadata so when you reference an image, settings will be pulled across automatically.
 
-We recommend using the [devcontainer CLI](/docs/devcontainers/devcontainer-cli.md) to pre-build your images since it is kept in sync with the Dev Containers extension's latest capabilities - including [dev container Features](#dev-container-features). Once you've built your image, you can push it to a container registry (like the [Azure Container Registry](https://learn.microsoft.com/azure/contai ...
+We recommend using the [dev container CLI](/docs/devcontainers/devcontainer-cli.md) to pre-build your images since it is kept in sync with the Dev Containers extension's latest capabilities - including [dev container Features](#dev-container-features). Once you've built your image, you can push it to a container registry (like the [Azure Container Registry](https://learn.microsoft.com/azure/conta ...
```
Maybe like this so that it is clear the Action can be used for image builds?  Certainly the easiest to do if you use Actions.

```suggestion
We recommend using the [Dev Container CLI](/docs/devcontainers/devcontainer-cli.md) (or other spec supporting utilities like the [GitHub Action](https://github.com/marketplace/actions/devcontainers-ci)) to pre-build your images since it is kept in sync with the Dev Containers extension's latest capabilities - including [dev container Features](#dev-container-features). Once you've built your image, you can push it to a container registry (like the [Azure Container Registry](https://learn.microsoft.com/azure/container-registry/container-registry-get-started-docker-cli?tabs=azure-cli), [GitHub Container Registry](https://docs.github.com/packages/working-with-a-github-packages-registry/working-with-the-container-registry#pushing-container-images), or [Docker Hub](https://docs.docker.com/engine/reference/commandline/push)) and reference it directly.
```

bamurtaugh replied:
Sounds good to me!

bamurtaugh commented:
@gregvanl This is ready for your review and merge whenever you're ready!

chrmarti commented on remote/advancedcontainers/sharing-git-credentials.md:
```diff
@@ -0,0 +1,104 @@
+
+# Sharing Git credentials with your container
+
+The Dev Containers extension provides out of box support for using local Git credentials from inside a container. In this section, we'll walk through the two supported options.
```
"out of the box"

bamurtaugh replied:
Fixed, thanks!

chrmarti reviewed (dismissed):
LGTM, 2 nits.

bamurtaugh commented on docs/devcontainers/containers.md:
```diff
@@ -366,6 +394,10 @@ If there are extensions that you would like always installed in any container, y
 
+### Avoiding extension reinstalls
+
+By default, VS Code will install extensions and VS Code Server inside the container's filesystem. While this has performance benefits over a locally mounted filesystem, the disadvantage is that VS Code will have to reinstall them on a container rebuild. If you find yourself rebuilding frequently, you can enable `dev.containers.cacheVolume` in your VS Code settings. This setting controls whether a ...
```
Thanks for pointing this out. Since it's enabled by default, I don't think we need this section, so I've removed it. This will also help with a bit more brevity for this doc.
